src/Primitives.hs

```diff
@@ -151,65 +151,61 @@ primitiveColumn x@(XObj _ i t) ctx args =
     go = maybe err (\info -> (ctx, Right (XObj (Num IntTy (fromIntegral (infoColumn info))) i t)))
 
 primitiveImplements :: Primitive
-primitiveImplements xobj ctx [x@(XObj (Sym interface@(SymPath _ _) _) _ _), inner@(XObj (Sym impl@(SymPath prefixes name) _) info _)] =
-  let global = contextGlobalEnv ctx
-      def = lookupInEnv impl global
-   in maybe notFound' found def
+primitiveImplements _ ctx [x@(XObj (Sym interface@(SymPath _ _) _) _ _), (XObj (Sym (SymPath prefixes name) _) info _)] =
+  do
+    (maybeInterface, maybeImpl) <- pure ((lookupInEnv interface tyEnv), (lookupInEnv (SymPath modules name) global))
+    case (maybeInterface, maybeImpl) of
+      (_, Nothing) ->
+        if null modules
+          then pure (evalError ctx "Can't set the `implements` meta on a global definition before it is declared." info)
+          else updateMeta (Meta.stub (SymPath modules name)) ctx
+      (Nothing, Just (_, implBinder)) ->
+        (warn >> updateMeta implBinder ctx)
+      (Just (_, interfaceBinder), Just (_, implBinder)) ->
+        (registerIt interfaceBinder implBinder)
   where
+    global = contextGlobalEnv ctx
+    tyEnv = getTypeEnv . contextTypeEnv $ ctx
     (SymPath modules _) = consPath (union (contextPath ctx) prefixes) (SymPath [] name)
-    checkInterface =
-      let warn = do
-            emitWarning
-              ( "The interface " ++ show interface ++ " implemented by " ++ show impl
-                  ++ " at "
-                  ++ prettyInfoFromXObj xobj
-                  ++ " is not defined."
-                  ++ " Did you define it using `definterface`?"
-              )
-          tyEnv = getTypeEnv . contextTypeEnv $ ctx
-       in maybe warn (\_ -> pure ()) (lookupInEnv interface tyEnv)
-    -- If the implementation binding doesn't exist yet, set the implements
-    -- meta. This enables specifying a function as an implementation before
-    -- defining it.
-    --
-    -- This is only allowed for qualified bindings. Allowing forward declarations on global bindings would cause a loop in
-    -- primitiveMetaSet's lookup which is generic.
-    notFound' =
-      if null modules
-        then pure $ evalError ctx "Can't set the `implements` meta on a global definition before it is declared." info
-        else
-          ( checkInterface
-              >> primitiveMetaSet xobj ctx [inner, XObj (Str "implements") (Just dummyInfo) (Just StringTy), XObj (Lst [x]) (Just dummyInfo) (Just DynamicTy)]
-          )
-    found (_, Binder meta defobj) =
-      checkInterface
-        >> either registerError updateImpls (registerInInterface ctx defobj interface)
+    warn =
+      emitWarning
+        ( "The interface "
+            ++ show (getPath x)
+            ++ " is not defined."
+            ++ " Did you define it using `definterface`?"
+        )
+    registerIt :: Binder -> Binder -> IO (Context, Either EvalError XObj)
```

Comment:
defineIt? See note below.

Reply:
(below in the code, above in the list of comments... :S)

Reply:
Changed to `addToInterface` I think this is more explicit and accurate, wdyt?